### ISSUE TITLE
Support HTTPS for custom domains when load balancer is set up in CloudFlare

### ISSUE THREAD

**Issue by s-h-a-d-o-w** [caprover/caprover#523]
**Is your feature request related to a problem? Please describe.**
When trying to enable HTTPS for my custom domain, I get `1107 : Verification Failed.`.
I tried it out on dnschecker.org and it's propagated properly.

**Describe the solution you'd like**
On a basic level that one can use HTTPS even when there is no A record that points to an IP. (Since doing that is not possible when using CloudFlare's load balancer.)

Ideally, one would also be able to add a custom certificate, so the connection to CloudFlare [can be verified](https://www.viget.com/articles/heroku-cloudflare-the-right-way/).

**Describe alternatives you've considered**
In another issue, @githubsaturn mentions that there has to be an A record that points to an IP.
This is not possible because I have to direct it via CNAME to the load balancer's name.

**Comment by githubsaturn** [comment]
Domain verification is a service that can be turned off if user needs to. See this for details:

#413

**Comment by s-h-a-d-o-w** [comment]
Unfortunately, it still doesn't work:
```
1107 : Unexpected output when enabling SSL forapp.mydomain.com with ACME Certbot Saving
 debug log to /var/log/letsencrypt/letsencrypt.log Plugins selected: Authenticator webroot, Installer 
None Obtaining a new certificate Performing the following challenges: http-01 challenge for 
app.mydomain.com Using the webroot path /captain-webroot/app.mydomain.com for all unmatched 
domains. Waiting for verification... Cleaning up challenges Failed authorization procedure. 
app.mydomain.com (http-01): urn:ietf:params:acme:error:unauthorized :: The client lacks sufficient 
authorization :: Invalid response from https://app.mydomain.com/.well-known/acme-
challenge/wVqedheRzo1RN2ZGNNMzs9jtLZCEQ8ucXSsIVi_JsTI [2606:4700:30::681b:8143]: "
<!doctype html>\n<html>\n\n<head>\n <meta charset=utf-8>\n <meta content=\"width=device-
width,initial-scale=1.0,minimum-scale=1." [1m IMPORTANT NOTES: [0m - The following errors 
were reported by the server: Domain: app.mydomain.com Type: unauthorized Detail: Invalid response 
from https://app.mydomain.com/.well-known/acme-challenge/wVqedheRzo1RN2ZGNNMzs9jtLZCEQ8ucXSsIVi_JsTI [2606:4700:30::681b:8143]: "
<!doctype html>\n<html>\n\n<head>\n <meta charset=utf-8>\n <meta content=\"width=device-width,initial-scale=1.0,minimum-scale=1." 
To fix these errors, please make sure that your domain 
name was entered correctly and the DNS A/AAAA record(s) for that domain contain(s) the right IP 
address.
```

As for what it says in "to fix these errors" - routing for the same host name without HTTPS works. As in `my machine -> HTTPS -> cloudflare -> HTTP -> caprover`.

**Comment by githubsaturn** [comment]
Oh I see. In that case, of course, you cannot use CloudFlare as an http proxy.

You can only have 1 HTTPS termination end point. If you use CloudFlare, you can enable HTTPS on CloudFlare, but you cannot enable https on CapRover as well.

Let's Encrypt, needs to verify your website by creating a temporary file inside .well-known directory. If this directory is being blocked by CloudFlare, of course, let's Encrypt won't be able to verify your website.

**Comment by s-h-a-d-o-w** [comment]
Are you sure? Because [here](https://support.cloudflare.com/hc/en-us/articles/360024787372-End-to-end-HTTPS-with-Cloudflare-Part-1-conceptual-overview), multiple ways are described to make end-to-end encryption possible and all of them describe using a certificate that's different from the one cloudflare uses.

**Comment by githubsaturn** [comment]
It may work with an SSL certificate that's been acquired through other ways, but for Let's Encrypt to work, it needs to be able to access your .well-known directory. From the logs that you shared:
```
Invalid response from https://app.mydomain.com/.well-known/acme-
challenge/wVqedheRzo1RN2ZGNNMzs9jtLZCEQ8ucXSsIVi_JsTI
```
It appears that CloudFlare doesn't forward the request properly to CapRover server. You can confirm this by getting nginx logs while you're clicking the ENABLE HTTPS button on CapRover. You can get nginx logs by `docker service logs captain-nginx --follow --since 5m`

**Comment by s-h-a-d-o-w** [comment]
Thank you for your support! Because of your description, I stumbled across [this post on the CloudFlare forum](https://community.cloudflare.com/t/lets-encrypt-and-cloudflare-how-to-set/66442/3) which made me realize "Huh, I guess that's right - the point is to just generate the certificate for this machine. CloudFlare doesn't need to be involved."

I've managed to make it work but wonder whether this process can't be automated/simplified. Because it's **very** cumbersome and I really don't want to do this manually for all of my apps.

Anyway, a step by step guide just in case other people stumble across this issue:
- If you already added one, delete the CNAME that points to the load balancer and instead add an A record with the IP of one of the machines. (Also disable CloudFlare proxying by switching from orange to gray cloud)
- Enable HTTPS via the CapRover dashboard.
- Repeat the previous two steps with the other machines. (Of course one by one, since the A record can only point to one IP at a time)
- Revert to the CNAME record pointing to the load balancer and enable CloudFlare proxying.
- (Optional) Confirm that HTTPS and location based routing work. (Use a VPN to visit the site from different regions and check out the airport codes of `cf-ray` in the response)
- Set encryption mode to "Full" in CloudFlare's SSL/TLS settings and switch on "Always Use HTTPS" in the "Edge Certificates" tab.

**Comment by githubsaturn** [comment]
This works only for 3 months! Every 3 months Let's Encrypt needs to perform a certificate renewal. Your certificate will be invalid after 3 months and your site will stop working.

Having said that, skimming over the CloudFlare docs, they seem to have support for "Expired" SSL certs. i.e., there may be an option to ignore the fact that SSL certificate is expired. In that case, things might work.

**Comment by s-h-a-d-o-w** [comment]
Oh, damn... thanks!

Maybe I should just look into how to add a certificate generated by CloudFlare to the CapRover instance instead, since that would make that workaround unnecessary and IIRC, that has an expiry of many years.

**Comment by aniketpanjwani** [comment]
I also encountered this same problem, and @s-h-a-d-o-w is correct; you need to change the template.

Specifically, do the following: 

1. Create an Origin Certificate at Cloudflare.
2. Download/save locally the .key/.pem files.
3. SSH into your CapRover server, and cd into /captain/data/nginx-shared.
4. Either copy and paste the .key/.pem files into new files on the server, or just scp them onto the server. Execute "chmod 600" on each.
5. In your application on CapRover, in the template under HTTP Auth settings, scroll down to where it says "Listen 443". Change the two values of ssl_certificate and ssl_certificate_key to the filepaths in nginx-shared.
6. Change the SSL Encryption mode on Cloudflare to Full (Strict).

It would be nice to automate this, but I'm pretty new to CapRover, so I can neither speculate whether it's feasible/easy, nor whether it's within the scope of the project.
